Call ProcessNewBlock() asynchronously

TheBlueMatt opened the pull request:
This is essentially a rebase of #12934, though its actually greenfield to reduce diff a bit. Based on #16174 as without it the parallelism would be almost entirely useless. Note that this is draft as I haven't (yet) bothered doing the backgrounding of the block validation, its currently just a return value change and the net_processing changes to handle it.

TheBlueMatt committed:
Add a new peer state tracking class to reduce cs_main contention.

CNodeState was added for validation-state-tracking, and thus,
logically, was protected by cs_main. However, as it has grown to
include non-validation state (taking state from CNode), and as
we've reduced cs_main usage for other unrelated things, CNodeState
is left with lots of cs_main locking in net_processing.

In order to ease transition to something new, this adds only a
dummy CPeerState which is held as a reference for the duration of
message processing.

Note that moving things is somewhat tricky pre validation-thread as
a consistent lockorder must be kept - we can't take a lock on the
new cs_peerstate in anything that's called directly from
validation.

TheBlueMatt committed:
Make ProcessNewBlock return a future instead of an immediate bool

This prepares for making ProcessNewBlock processing actually happen
in a separate thread from the caller, even though all callsites
currently block on the return value immediately.

Note that this makes some of the unit tests less restrictive.

TheBlueMatt committed:
Actually parallelize validation somewhat

TheBlueMatt committed:
Move nDoS counters to CPeerState (and, thus, out of cs_main)

TheBlueMatt committed:
Move rejects into cs_peerstate.

This removes the cs_main lock which is taken on every
ProcessMessages call.

TheBlueMatt committed:
Run ProcessNewBlock in a background thread (XXX: p2p_sendheaders broken)

Spawn a background thread at startup which validates each block as
it comes in from ProcessNewBlock, taking advantage of the new
std::future return value to keep tests simple (and the new
net_processing handling of such values async already).

This makes introducing subtle validationinterface deadlocks much
harder as any locks held going into ProcessNewBlock do not interact
with (in the form of lockorder restrictions) locks taken in
validationinterface callbacks.

Note that after this commit, feature_block and feature_assumevalid
tests time out due to increased latency between block processing
when those blocks do not represent a new best block. This will be
resolved in the next commit.

TheBlueMatt committed:
Move BlockChecked to a background thread

TheBlueMatt commented:
Oops, it turns out its actually not possible to do #16174 first separately (and without it, this PR is entirely useless). By itself, #16174 ends up hitting the deadly cross-validationinterface-queue deadlock - it takes cs_peerstate first, then (eventually) calls PNB, which may call LimitValidationInterfaceQueue(). Meanwhile, in the validation queue, we may end up waiting on cs_peerstate to update some peer's state. Of course, once we actually move block validation into its own thread, this is no longer (materially) a concern - we trivially guarantee there are no locks held going into LimitValidationInterfaceQueue().

TheBlueMatt commented:
Note that there is still a regression causing some functional tests to time out as we may end up waiting 100ms to discover that a block has been processed and we can process the next message from that peer (this shouldn't be an issue in the case of downloading from multiple peers since we'll always wake on a new tip block, and seems to not be an issue for one peer, even though its possible it would be due to a wake race). Not 100% sure what the best solution to that is.

jamesob commented:
> Oops, it turns out its actually not possible to do #16174 first separately [...] Meanwhile, in the validation queue, we may end up waiting on cs_peerstate to update some peer's state.

For other reviewers: the particular validationinterface callback at fault here looks to be `PeerLogicValidation::BlockChecked`.

Big Concept ACK on the lock-splitting at the very least, though I do wish it could be done separately somehow. I'm a Concept-Concept ACK on this (in concept), but want to make sure I fully understand the implications.

TheBlueMatt committed:
Add a callback to indicate a block has been processed

This resolves the performance regression introduced in the previous
commit by always waking the message processing thread after each
block future resolves.

Sadly, this is somewhat awkward - all other validationinterface
callbacks represent an actual change to the global validation state,
whereas this callback indicates only that a call which one
validation "client" made has completed. After going back and forth
for some time I didn't see a materially better way to resolve this
issue, and luckily its a rather simple change, but its far from
ideal. Note that because we absolutely do not want to ever block on
a ProcessNewBlock-returned-future, the callback approach is
critical.

TheBlueMatt committed:
Move net_processing's ProcessNewBlock calls to resolve async.

Essentially, our goal is to not process anything for the given peer
until the block finishes processing (emulating the previous behavior)
without actually blocking the ProcessMessages loops. Obviously, in
most cases, we'll just go on to the next peer and immediately hit a
cs_main lock, blocking us anyway, but this we can slowly improve
that state over time by moving things from CNodeState to CPeerState.

TheBlueMatt committed:
Remove unnecessary cs_mains in denialofservice_tests

9fdf05d70cac4a62d1aeeb4299e2c3a9a866f8af resolved some lock
inversion warnings in denialofservice_tests, but left in a number
of cs_main locks that are unnecessary (introducing lock inversion
warnings in future changes).

practicalswift commented on src/validation.h:
```diff
@@ -208,15 +209,17 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
  * @param[in]   fForceProcessing Process this block even if unrequested; used for non-network block sources and whitelisted peers.
- * @param[out]  fNewBlock A boolean which is set to indicate if the block was first received via this call
- * @return True if state.IsValid()
+ * @return      A future which complets with a boolean which is set to indicate if the block was first received via this call
```
Adding an "e" completes the spelling here :-)

ryanofsky commented on src/validation.cpp:
```diff
@@ -3430,48 +3430,87 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
+            bool ret = CheckBlock(*pblock, state, chainparams.GetConsensus());
+            if (ret) {
+                // Store to disk
+                ret = ::ChainstateActive().AcceptBlock(pblock, state, chainparams, &pindex, fForceProcessing, nullptr, &fNewBlock);
```
In commit "Run ProcessNewBlock in a background thread (XXX: p2p_sendheaders broken)" (64e74ce8a917e071b8a250a4ba9cd110448796d6)

Throughout this function shouldn't `::ChainstateActive()` be replaced by `*this`?

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -745,12 +746,13 @@ static UniValue submitblock(const JSONRPCRequest& request)
-    if (!new_block && accepted) {
+    LOCK(cs_main);
+    const CBlockIndex* pindex = LookupBlockIndex(blockptr->GetHash());
+    if (!new_block && pindex && pindex->IsValid()) {
```
In commit "Make ProcessNewBlock return a future instead of an immediate bool" (79e6e446dd0c8a3f53759a9792388f28e1053bfb)

This seems fine, but just to make sure I understand correctly, there is a change in behavior here? Previously `accepted` would only be true if `CheckBlock` and `AcceptBlock` and `ActivateBestChain` all succeeded but this can be true even they fail?

ryanofsky commented on src/test/blockfilter_index_tests.cpp:
```diff
@@ -175,7 +175,7 @@ BOOST_FIXTURE_TEST_CASE(blockfilter_index_initial_sync, TestChain100Setup)
     uint256 chainA_last_header = last_header;
     for (size_t i = 0; i < 2; i++) {
         const auto& block = chainA[i];
-        BOOST_REQUIRE(ProcessNewBlock(Params(), block, true, nullptr));
```
In commit "Make ProcessNewBlock return a future instead of an immediate bool" (79e6e446dd0c8a3f53759a9792388f28e1053bfb)

In the commit message about these tests you wrote "Note that this makes some of the unit tests less restrictive." But I don't think there's a good justification for doing this. If you just declared `struct ProcessNewBlockResult { bool error = false; bool already_have = false; };` and returned `std::future<ProcessNewBlockResult>`, the tests could behave the same and the code would be more readable and less fragile, without the need for people to guess from context whether the bool indicates success or newness.

ajtowns replied:
Think I agree with this suggestion; dropping the `fNewBlock` result at the same time as turning the return into a `future` seems unnecessarily complicated.

ryanofsky reviewed (commented):
Started review (will update list below with progress).
- [X] 9bc8ca61f5c9ca57ea71b3acdd1021a5cf66512f Remove unnecessary cs_mains in denialofservice_tests (1/10)
- [X] 79e6e446dd0c8a3f53759a9792388f28e1053bfb Make ProcessNewBlock return a future instead of an immediate bool (2/10)
- [X] e89822964e9ef8dadf88d47bc94f5396a8c61a54 Add a new peer state tracking class to reduce cs_main contention. (3/10)
- [ ] 8ddbb126780a8ab6ff547411e2b8d3b46e42c887 Move net_processing's ProcessNewBlock calls to resolve async. (4/10)
- [ ] 64e74ce8a917e071b8a250a4ba9cd110448796d6 Run ProcessNewBlock in a background thread (XXX: p2p_sendheaders broken) (5/10)
- [ ] ba810b9e88bee558c34ac24dc685c68e5d5c83cb Add a callback to indicate a block has been processed (6/10)
- [ ] 3499d988a9cef162987d98dda3cf914be24a3fa1 Move BlockChecked to a background thread (7/10)
- [ ] 445bb84bd3c38b2e98ccc247aa1a346ca3c5d66c Move nDoS counters to CPeerState (and, thus, out of cs_main) (8/10)
- [ ] b6a05524f9b66af1147c9776e3b499bdb94655f8 Move rejects into cs_peerstate. (9/10)
- [ ] 2d1e8f64cd527b9c44ba171beb0bb837c64cc0a5 Actually parallelize validation somewhat (10/10)

jamesob commented:
When IBDing from a single peer, this branch appears to be no slower (or negligibly slower) than master. I'm going to modify the benchmark framework to test against multiple peers, which is where I'm assuming we'd expect to see some speedup.

![ibd local range 500000 505000](https://user-images.githubusercontent.com/73197/59466913-65c20580-8dfc-11e9-9c29-514f943a3a29.png)

ryanofsky commented:
Just want to note that all my comments above are just suggested cleanups. Feel free to ignore them if they don't make sense or aren't worth effort to implement. The only thing I'd really like to see are more comments about locking. When a lock is held in a small scope for specific purpose, I don't think there's a need to have a comment, but when a lock is held over wide scope for no obvious reason it's good to have a comment that suggests why it needs to be acquired there and when it is safe to release.

TheBlueMatt commented:
Working on rewriting this to make it simpler but it looks like it may end up growing so dunno about it.
